File: PROLOG/libreria.pl
escribio(elsaBornemann, socorro).
escribio(neilGaiman, sandman).
escribio(alanMoore, watchmen).
escribio(neilGaiman, americanGods).
escribio(neilGaiman, buensoPresagios).
escribio(terryPratchett, buensoPresagios).
escribio(brianAzarello, cienBalas).
escribio(warenElis, planetary).
escribio(frankMiller, elCaballeroOscuroRegresa).
escribio(frankMiller, batmanAnioUno).
escribio(isaacAsimov, fundacion).
escribio(isaacAsimov, yoRobot).
escribio(isaacAsimov, elFinDeLaEternidad).
escribio(isaacAsimov, laBusquedaDeLosElementos).
escribio(joseHernandez, martinFierro).
escribio(stephenKing, it).
escribio(stephenKing, misery).
escribio(stephenKing, carrie).
escribio(stephenKing, elJuegoDeGerald).
escribio(julioCortazar, rayuela).
escribio(jorgeLuisBorges, ficciones).
escribio(jorgeLuisBorges, elAleph).
escribio(horacioQuiroga, cuentosDeLaSelva).
escribio(horacioQuiroga, cuentosDeLocuraAmorYMuerte).

% ¿Es cierto que alguien escribió una determinada obra?
alguienEscribio(Obra) :- escribio(_, Obra).

% ¿Quién o quienes escribieron una obra?
quienesEscribieron(Alguien, Obra) :- escribio(Alguien, Obra).

% ¿Qué obra escribió cierta persona?
queObraEscribio(Alguien, Obra) :- escribio(Alguien, Obra).

% Si es cierto que cierta persona escribió alguna obra, sin importar cual.
esEscritor(Alguien) :- escribio(Alguien, _).

% Si es cierto que cierta obra existe.
existeObra(Obra) :- escribio(_, Obra).

esComic(sandman).
esComic(cienBalas).
esComic(watchmen).
esComic(planetary).
esComic(elCaballeroOscuroRegresa).
esComic(batmanAnioUno).

esArtistaDeNovenoArte(Artista) :- escribio(Artista, Obra), esComic(Obra).

odia(platon, diogenes).

numero(1).
numero(2).
numero(3).
numero(4).
numero(5).

% siguienteNoInversible(N,S) :- S is N + 1.
siguiente(N,S) :- numero(N), S is N + 1.

programaEn(maria, cobol).
programaEn(maria, java).
programaEn(mario, cobol).
programaEn(mario, python).
programaEn(jose, cobol).
programaEn(jorge,java).
programaEn(jorge,python).

% ¿Es cierto que alguien programa en Cobol?
alguienProgramaEn(Programa) :- programaEn(_, Programa).

% ¿Quién o quienes programan en Cobol?
quienesProgramanEn(Alguien, Programa) :- programaEn(Alguien, Programa).

% ¿Qué lenguajes maneja María?
enQuePrograma(Alguien, Programa) :- programaEn(Alguien, Programa).

% ¿Maria y Mario son colegas?
sonColegas(P1, P2) :- programaEn(P1, Programa), programaEn(P2, Programa).

% ¿Quienes son colegas de lenguajes?
colegas(P1, P2) :- programaEn(P1, Programa), programaEn(P2, Programa), P1 @< P2.

esHijoDe(carlosIII, isabelII).
esHijoDe(ana, isabelII).
esHijoDe(andrew, isabelII).
esHijoDe(edward, isabelII).
esHijoDe(louise, edward).
esHijoDe(james, edward).
esHijoDe(eugenia, andrew).
esHijoDe(beatriz, andrew).
esHijoDe(sienna, beatriz).
esHijoDe(august, eugenia).
esHijoDe(zara, ana).
esHijoDe(peter, ana).
esHijoDe(enrique, carlosIII).
esHijoDe(guillermo, carlosIII).
esHijoDe(jorge, guillermo).
esHijoDe(carlota, guillermo).
esHijoDe(luis, guillermo).
esHijoDe(sabannah, peter).
esHijoDe(isla, peter).
esHijoDe(lucas, zara).
esHijoDe(lena, zara).
esHijoDe(mia, peter).
esHijoDe(archie, enrique).
esHijoDe(lilibeth, enrique).

% ¿Es cierto que la reina Isabel tiene hijos?	
% esHijoDe(_,isabelII).

% ¿Quién o quienes son hijos de la reina Isabel?
% esHijoDe(Hijo,isabelII).

% ¿Quiénes son los nietos del Carlos III?
% esNietoDe(carlosIII, Nieto).

esNietoDe(Persona, Nieto) :- esHijoDe(Hijo, Persona), esHijoDe(Nieto, Hijo).

% ¿Es cierto que Archie es primo del príncipe Jorge?
% esPrimoDe(archie, jorge).
% 
% ¿Quienes son primos?
% esPrimoDe(P1, P2).

sonHermanos(P1, P2):- esHijoDe(P1, Padre), esHijoDe(P2, Padre).

esPrimoDe(P1, P2) :- esHijoDe(P1, Padre1), esHijoDe(P2, Padre2), P1 @< P2, sonHermanos(Padre1, Padre2), Padre1 \= Padre2.

No se tiene inversibilidad si:
Hechos con variables.
Comparación por distinto
Negación (dado que no liga variables)
>, >=, <, <=
Is/2

File: sol/libreria.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Eta reduce" #-}
{-# HLINT ignore "Use null" #-}

data Libro = Libro {
  autor:: String,
  obra:: String
} deriving (Show, Eq)

libros :: [Libro]
libros = [
  Libro "elsaBornemann" "socorro",
  Libro "neilGaiman" "sandman",
  Libro "alanMoore" "watchmen",
  Libro "neilGaiman" "americanGods",
  Libro "neilGaiman" "buenosPresagios",
  Libro "terryPratchett" "buenosPresagios",
  Libro "brianAzarello" "cienBalas",
  Libro "warenElis" "planetary",
  Libro "frankMiller" "elCaballeroOscuroRegresa",
  Libro "stephenKing" "elCaballeroOscuroRegresa",
  Libro "frankMiller" "batmanAnioUno",
  Libro "isaacAsimov" "fundacion",
  Libro "isaacAsimov" "yoRobot",
  Libro "isaacAsimov" "elFinDeLaEternidad",
  Libro "isaacAsimov" "laBusquedaDeLosElementos",
  Libro "joseHernandez" "martinFierro",
  Libro "stephenKing" "it",
  Libro "stephenKing" "misery",
  Libro "stephenKing" "carrie",
  Libro "stephenKing" "elJuegoDeGerald",
  Libro "julioCortazar" "rayuela",
  Libro "jorgeLuisBorges" "ficciones",
  Libro "jorgeLuisBorges" "elAleph",
  Libro "horacioQuiroga" "cuentosDeLaSelva",
  Libro "horacioQuiroga" "cuentosDeLocuraAmorYMuerte"
  ]

-- condicionEnLista condicion lista parametro = condicion (\libro -> obra libro == parametro) lista

-- ¿Es cierto que alguien escribió una determinada obra?
escribio:: String -> Bool
-- escribio obra = any (tituloCoincide obra) libros
escribio obraConsultar = any (\libro -> obra libro == obraConsultar) libros
-- escribio obraConsultar = condicionEnLista any libros obraConsultar

tituloCoincide :: String -> Libro -> Bool
tituloCoincide obraConsultar libro = obraConsultar == obra libro

-- ¿Quién o quienes escribieron una obra?
quienEscribio :: String -> [String]
quienEscribio obraConsultar = map autor (filter (\libro -> obra libro == obraConsultar) libros)

-- ¿Qué obra/s escribió cierta persona?
queEscribio :: String -> [String]
queEscribio autorConsultar = map obra (filter (\libro -> autor libro == autorConsultar) libros)

-- Si es cierto que cierta persona escribió alguna obra, sin importar cual.
esEscritor :: String -> Bool
esEscritor autorConsultar = any (\libro -> autor libro == autorConsultar) libros

-- Si es cierto que cierta obra existe.
existe :: String -> Bool
existe obraConsultar = any (\libro -> obra libro == obraConsultar) libros

-- //// FORMA DISTINTA DE HACERLO ;) ///////////////////////////////////////////////////

type Obra = String
type Nombre = String

data Autor = Autor {
  nombre:: Nombre,
  obras:: [Obra]
} deriving (Show, Eq)

autores:: [Autor]
autores = [
  Autor "elsaBornemann" ["socorro"],
  Autor "neilGaiman" ["sandman"],
  Autor "alanMoore" ["watchmen"],
  Autor "neilGaiman" ["americanGods","buenosPresagios"],
  Autor "terryPratchett" ["buenosPresagios"],
  Autor "brianAzarello" ["cienBalas"],
  Autor "warenElis" ["planetary"],
  Autor "frankMiller" ["elCaballeroOscuroRegresa", "batmanAnioUno"],
  Autor "isaacAsimov" ["fundacion", "yoRobot", "elFinDeLaEternidad", "laBusquedaDeLosElementos"],
  Autor "joseHernandez" ["martinFierro"],
  Autor "stephenKing" ["it", "misery", "carrie", "elJuegoDeGerald"],
  Autor "julioCortazar" ["rayuela"],
  Autor "jorgeLuisBorges" ["ficciones", "elAleph"],
  Autor "horacioQuiroga" ["cuentosDeLaSelva", "cuentosDeLocuraAmorYMuerte"] 
  ]

-- Quien o quienes escribieron determinada obra?
quienesEscribieron :: Obra -> [Nombre]
quienesEscribieron obraBuscada = map nombre (filter (escribioObra obraBuscada) autores)

escribioObra :: Obra -> Autor -> Bool
escribioObra obraBuscada autor = obraBuscada `elem` obras autor

-- Es cierto que alguien escribió determinada obra?
alguienEscribio :: Obra -> Bool
alguienEscribio obraBuscada = length (quienesEscribieron obraBuscada) > 0

-- Qué obras escribió determinado autor?
buscarPorNombre :: Nombre -> [Autor] -> Maybe Autor
buscarPorNombre _ [] = Nothing
buscarPorNombre nombreBuscado (x:xs)
    | nombre x == nombreBuscado = Just x
    | otherwise = buscarPorNombre nombreBuscado xs

obrasDeAutor :: Nombre -> [Obra]
obrasDeAutor autorBuscado = 
  case buscarPorNombre autorBuscado autores of
    Just autor -> obras autor
    Nothing -> error "No existe dicho Autor"

-- Si es cierto que cierta persona escribó alguna obra, sin importar cual 
escribioUnaObra :: Nombre -> Bool
escribioUnaObra autorBuscado = any (\autor -> nombre autor == autorBuscado && length (obras autor) > 0 ) autores

-- Si es cierto que cierta obra existe
existeObra :: Obra -> Bool
existeObra obraBuscada = any (\autor -> obraBuscada `elem` obras autor) autores
